### ISSUE TITLE
refactor documentation to remove react-desc

### ISSUE THREAD

**Issue by ericsoderberghp** [grommet/grommet#2976]
We can have issues in our documentation due to `react-desc` trying to combine `propTypes` with documentation and not always working well on either side. For example, we cannot reference propTypes from react-desc in other components due to how they are decorated. And, we need the ability to provide richer example syntaxes that don't work with the more simplified propType derived approach.

A suggestion is to have two files per component: `propTypes.js` and `doc.js`. The former would just have the `propTypes` and could be reference from other components. The latter would change the current documentation by removing the programatic prop definitions and instead support a richer `syntaxExamples` per each property, similar to what we do in `grommet-site`. The README.md files per component would be generated from `doc.js` with better syntax examples that aligned with grommet-site.

**Comment by oorestisime** [comment]
How about we just write the proptypes in the component directly?
Should remove some friction of working on grommet (having to open multiple files for the same component currently the X.js, doc.js, index.d.ts and maybe propTypes.js)

**Comment by atanasster** [comment]
Great to see some thiught given finally to take out react-desc. It has been truly a problem for code completion, debugging and overall complexity.

I agree the propTypes should be in the component file and remove the complexities introduced with doc.js files because of react-desc once and for all.

However i suggest completely removing react-desc - not just for prop types, but also for documenting. There are much better choices out there like styleguidist, docz etc, that would also support a future move to typescript.

**Comment by harunurhan** [comment]
Having only `index.d.ts`, even going further by having `TypeScript` typings inside the component files (does not require migrating whole codebase to `TypeScript`) would be much better, IMHO.

Because:
- `TypeScript` is more expressive and powerful when it comes to defining types, compared to `prop-types`, even to `flow`.
- It has tooling for generating docs.
- You get build time static type checking which will help to catch bugs earlier.
- Will solve problems that are caused by inconsistent typings (different types for `prop-types` and `TypeScript`).

**Comment by ericsoderberghp** [comment]
Two years later ...

- I think we should remove `react-desc` entirely.
- I think we should out the propTypes in a separate file, that makes the typescript typings and prop types consistent. I suggest this because I think there is enough complexity in some of the components typings that adding them into the component file would make it a bit unwieldy.
- I think all of the documentation should be in moved to the grommet-site repository. This will allow more customization of how complex types are described but just describing them in markup, rather than building a more complex JSON structure just to then have to convert that to markup later.
- I think we should update grommet-site to show properties with different types where there are different descriptions and examples for each type, rather than trying to merge those into a single description.

**Comment by jcfilben** [comment]
If I am understanding correctly, the `doc.js` files are being completely removed, the `index.d.ts` files are not being changed and a new file is being added, `propTypes.js` that has the propTypes that were originally in the `doc.js` file?

> I think we should remove react-desc entirely.

isn't PropTypes coming from react-desc?

**Comment by ericsoderberghp** [comment]
> If I am understanding correctly, the `doc.js` files are being removed from Grommet, the `index.d.ts` files are not being changed and a new file is being added, `propTypes.js` that has the propTypes that were originally in the `doc.js` file?

yes

> > I think we should remove react-desc entirely.
> 
> isn't PropTypes coming from react-desc?

It is today. The proposal is to add a propTypes.js that will just contain the `prop-types` stuff. https://github.com/facebook/prop-types#readme

**Comment by jcfilben** [comment]
Update:
I created a pull request on both the grommet-site and grommet repo for the Anchor component showing the steps needed to remove react desc and move the documentation to the grommet-site. I opened a separate issue [here](#5353) to track which components need to be changed, and outlined the steps to do so in both repos. The new issue already contains a lot of information which is why I didn't just add on to this issue. Also since this issue is over 2 years old, I though a new issue might be easier for people to find and help get more contributions.

We can close this issue in favor of #5353 or leave it open until react-desc has been completely removed
